src/orion/algo/tpe.py

```diff
@@ -0,0 +1,351 @@
+            below_points = numpy.array([flatten_dims(point, self.space) for point in below_points])
+            above_points = numpy.array([flatten_dims(point, self.space) for point in above_points])
+            idx = 0
+            for dimension in self.space.values():
+
+                if dimension.type == 'real':
+                    shape = dimension.shape
+                    if not shape:
+                        shape = (1,)
+                    # Unpack dimension
+                    for j in range(shape[0]):
+                        idx = idx + j
+                        new_point = self._sample_real(dimension,
+                                                      below_points[:, idx],
+                                                      above_points[:, idx])
+                        point.append(new_point)
+
+                else:
+                    raise ValueError("TPE now only support Real Dimension.")
+
+                idx += 1
+            point = regroup_dims(point, self.space)
+            samples.append(point)
+        return samples
+
+    def _sample_real(self, dimension, below_points, above_points):
+        """Return a real point based on the observed good and bad points"""
+        low, high = dimension.interval()
+        below_mus, below_sigmas, blow_weights = \
+            adaptive_parzen_estimator(below_points, low, high, self.prior_weight,
+                                      self.equal_weight, flat_num=25)
+        above_mus, above_sigmas, above_weights = \
+            adaptive_parzen_estimator(above_points, low, high, self.prior_weight,
+                                      self.equal_weight, flat_num=25)
+
+        gmm_sampler_below = GMMSampler(self, below_mus, below_sigmas, low, high, blow_weights)
+        gmm_sampler_above = GMMSampler(self, above_mus, above_sigmas, low, high, above_weights)
+
+        points = gmm_sampler_below.sample(self.n_ei_candidates)
+        lik_blow = gmm_sampler_below.get_loglikelis(points)
+        lik_above = gmm_sampler_above.get_loglikelis(points)
+
+        new_point = compute_max_ei_point(points, lik_blow, lik_above)
+
+        return new_point
+
+    def split_trials(self):
+        """Split the observed trials into good and bad ones based on the ratio `gamma``"""
+        sorted_trials = sorted(self._trials_info.values(), key=lambda x: x[1]['objective'])
+        sorted_points = [list(points) for points, results in sorted_trials]
+
+        split_index = int(numpy.ceil(self.gamma * len(sorted_points)))
+
+        below = sorted_points[:split_index]
+        above = sorted_points[split_index:]
+
+        return below, above
+
+    def observe(self, points, results):
+        """Observe evaluation `results` corresponding to list of `points` in
+        space.
+
+        A simple random sampler though does not take anything into account.
+        """
+        super(TPE, self).observe(points, results)
+
+
+class GMMSampler():
```

Comment:
I think it would be better to use [sklearn's Gaussian Mixture](https://scikit-learn.org/stable/modules/generated/sklearn.mixture.GaussianMixture.html#sklearn.mixture.GaussianMixture). This way we could also easily swap it to kernel [density estimator (parzen window)](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KernelDensity.html#sklearn.neighbors.KernelDensity).

Reply:
Forget about it. Ready more carefully the details on the Parzen estimator in the original paper I understand that we could hardly do this with implementations in sklearn. That's fine then.